Add support for (Atom) feeds

frankieroberto opened the pull request:
Adds a new template for [Atom feeds](https://en.wikipedia.org/wiki/Atom_(web_standard)), allowing people to subscribe to receive content within a feed reader.

paulrobertlloyd commented on layouts/base.njk:
```diff
@@ -45,6 +45,8 @@
   <link rel="stylesheet" href="{{ stylesheet | canonicalUrl }}">
   {% endfor %}
 
+  {% if options.feedUrl %}<link rel="alternate" type="application/atom+xml" href="{{ options.feedUrl }}" >{% endif %}
```
Add canonical URL so that if a relative path is provided it gets properly resolved.

```suggestion
  {% if options.feedUrl %}<link rel="alternate" type="application/atom+xml" href="{{ options.feedUrl | canonicalUrl }}" >{% endif %}
```

paulrobertlloyd commented:
Looks good. Need to add documentation for this new option, too.

Am also wondering if the template added in https://github.com/x-govuk/x-govuk.github.io/pull/109 should instead be upstreamed into this project, with guidance similar to that we have for [adding a site search](https://x-govuk.github.io/govuk-eleventy-plugin/search/).

Could take a collection to include in the feed:

```markdown
---
eleventyExcludeFromCollections: true
layout: feed-xml
permalink: /posts.xml
collection: posts
---
```

frankieroberto commented:
@paulrobertlloyd @stevenjmesser ok I've upstream the Atom feed layout into this plugin, and added some initial docs. Let me know what you think.

I've set up a feed for the docs site itself for testing purposes, but given that the docs site doesn’t have any (chronological) posts, it’s empty.

paulrobertlloyd commented on docs/feed.md:
````diff
@@ -0,0 +1,39 @@
+```js
+const govukEleventyPlugin = require('@x-govuk/govuk-eleventy-plugin')
+
+eleventyConfig.addPlugin(govukEleventyPlugin, {
+  feedUrl: 'https://your-website.example/feed.xml',
+  [...]
+})
+```
````
Elsewhere in the documentation, we only show the property that need to be added.

```suggestion
eleventyConfig.addPlugin(govukEleventyPlugin, {
  feedUrl: 'https://your-website.example/feed.xml'
})
```
```

paulrobertlloyd commented on eleventy.config.js:
```diff
@@ -20,6 +23,7 @@ module.exports = function (eleventyConfig) {
     url:
       process.env.GITHUB_ACTIONS &&
       'https://x-govuk.github.io/govuk-eleventy-plugin/',
+    feedUrl: 'https://x-govuk.github.io/govuk-eleventy-plugin/feed.xml',
     stylesheets: ['/assets/application.css'],
```
Not sure the documentation site needs a feed, and the feed doesn’t include anything, anyway.

```suggestion
      'https://x-govuk.github.io/govuk-eleventy-plugin/',
    stylesheets: ['/assets/application.css'],
```

frankieroberto replied:
Agree. 👍

paulrobertlloyd commented on docs/feed.njk:
```diff
@@ -0,0 +1,5 @@
+---
+eleventyExcludeFromCollections: true
+layout: feed
+permalink: /feed.xml
+---
```
Not sure we need this for the documentation site.

paulrobertlloyd commented on layouts/feed.njk:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Looks correct, but have we tested this works upstream in for posts on the X-GOVUK website?

frankieroberto replied:
I’ve tested this by running `npm pack` and then copying the zip file to the x-govuk.github.io repo, and referencing it there from the `package.json`. Seems to work!

paulrobertlloyd reviewed (changes requested):
Let’s remove the empty feed. I agree it’s useful for testing, but I think if it becomes an issue, we should a separate test build (in CI?) rather than use the documentation site for testing as well.

frankieroberto commented:
@paulrobertlloyd re-review please! I've updated the documentation to also mention creating the `post` collection in the config file. You could do this with tags, but we’ve moved away from that to avoid every post having a "post" tag, which is annoying if you display the tags.

paulrobertlloyd commented on docs/feed.md:
````diff
@@ -21,6 +21,18 @@ permalink: /feed.xml
 
 The `permalink` value is the location of the generated feed.
 
+## Create a collection of posts
+
+The feed will include all pages in a collection called 'post'.
+
+You can create this by adding some code to your `eleventy.config.js`:
+
+```js
+eleventyConfig.addCollection('post', (collection) => {
+  return collection.getFilteredByGlob('app/posts/*.md')
+})
+```
+
````
I wonder if it’s better to be a little more explicit, and call this collection `feed` (the Atom feed might not contain what the author considers to be ‘posts’).

Also, I wonder if it’s possible to create a default collection, that could be overridden. Maybe a `feed` collection that gets all Markdown files with an explicit date (`page.data.date` I think)? Maybe that’s something to think about for later.

paulrobertlloyd replied:
@frankieroberto Thoughts on the above?

frankieroberto replied:
Sorry, missed this. I like `feed` as a default collection name.

Can you remind me if there’s a way for the page itself to set the collection name to override the default collection referenced in the template? (That way you could have multiple feeds in a project if needed)

paulrobertlloyd replied:
Without testing, you could do something like this in the feed template:

```njk
{% set collection = data.collection or "feed" %}

{%- for post in collections[collection] | reverse %}
  …
{%- endfor %}
```

Then, in the front matter, a user could do the following:

```yaml
---
eleventyExcludeFromCollections: true
layout: feed
permalink: /feed.xml
collection: kittens
---
```

frankieroberto replied:
@paulrobertlloyd I'm wondering now whether it makes sense to use the pagination options, in the same way as the `collection` layout?

Then you could do:

```yaml
eleventyExcludeFromCollections: true
layout: feed
permalink: /feed.xml
pagination:
  data: collections.post
  size: 20
  reverse: true
```

And in the layout itself:

```nunjucks
{%- for item in pagination.items %}
```

paulrobertlloyd replied:
Oh yes, yes, yes! This is precisely the right approach.

paulrobertlloyd commented on docs/feed.md:
```diff
@@ -0,0 +1,50 @@
+const govukEleventyPlugin = require('@x-govuk/govuk-eleventy-plugin')
+
+eleventyConfig.addPlugin(govukEleventyPlugin, {
+  feedUrl: 'https://your-website.example/feed.xml'
```
Given the use of the `canonicalUrl` filter, I think you can use this instead (saves issues with GitHub Pages URLs possibly being in sub-folder, etc.) 

```suggestion
  feedUrl: 'feed.xml'
```

frankieroberto commented:
@paulrobertlloyd I think this is good to merge now? Then I'll do a release.

paulrobertlloyd reviewed (approved):
Perfect, thanks for adding this. Let’s release as v6.2.0
